File: Implementacion C#/MatrixProductProject/MatrixProductMemParallel.cs
using System;
using System.Threading.Tasks;

namespace MatrixProductNamespace
{
    public class MatrixProductMemParallel
    {
        public static void ParallelProduct(int matrixSize)
        {
            // Establecer el número de hilos de OpenBLAS
            MatrixProduct.openblas_set_num_threads(Environment.ProcessorCount);

            double[] MatrixA = MatrixProductMemory.GenerateRandomMatrix(matrixSize, matrixSize);
            double[] MatrixB = MatrixProductMemory.GenerateRandomMatrix(matrixSize, matrixSize);

            long startTime = DateTime.Now.Ticks;

            double[] MatrixC = new double[matrixSize * matrixSize];

            int numThreads = Environment.ProcessorCount;
            int blockSize = (matrixSize + numThreads - 1) / numThreads;

            Parallel.For(0, numThreads, i =>
            {
                int startRow = i * blockSize;
                int endRow = Math.Min(startRow + blockSize, matrixSize);
                int rowCount = endRow - startRow;

                double[] subMatrixA = new double[rowCount * matrixSize];
                double[] subMatrixC = new double[rowCount * matrixSize];

                Array.Copy(MatrixA, startRow * matrixSize, subMatrixA, 0, rowCount * matrixSize);

                MatrixProduct.cblas_dgemm(101, 111, 111,
                    rowCount, matrixSize, matrixSize,
                    1.0, subMatrixA, matrixSize,
                    MatrixB, matrixSize,
                    0.0, subMatrixC, matrixSize);

                Array.Copy(subMatrixC, 0, MatrixC, startRow * matrixSize, rowCount * matrixSize);
            });

            long endTime = DateTime.Now.Ticks;

            Console.WriteLine("Product completed. First 10 columns in first row:");
            for (int i = 0; i < 10; i++)
            {
                Console.Write(i < 9
                    ? $" {MatrixC[i]:F2} , "
                    : $" {MatrixC[i]:F2} \n");
            }
            Console.WriteLine($"Product completed in {(endTime - startTime) / 10000000.0:F2} seconds.");
        }
    }
}

File: Implementacion Java/Matrix Product JBlas/matrixproduct/src/main/java/com/afflorezc/MatrixProductMemParallel.java
package com.afflorezc;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import org.jblas.DoubleMatrix;

public class MatrixProductMemParallel {

    static int numThreads = 4;
    static DoubleMatrix matrixA;
    static DoubleMatrix matrixB;
    static DoubleMatrix matrixC;

    public static void parallelProduct(int matrixSize) {

        matrixA = DoubleMatrix.rand(matrixSize, matrixSize);
        matrixB = DoubleMatrix.rand(matrixSize, matrixSize);

        long startTime = System.nanoTime();

        matrixC = new DoubleMatrix(matrixSize, matrixSize);

        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
        int blockSize = matrixSize / numThreads;

        for (int i = 0; i < numThreads; i++) {
            int startRow = i * blockSize;
            int endRow = (i == numThreads - 1) ? matrixSize : (i + 1) * blockSize;
            executor.submit(new MatrixMultiplier(matrixSize, startRow, endRow));
        }

        executor.shutdown();
        while (!executor.isTerminated()) {}

        long endTime = System.nanoTime();

        System.out.println("Product completed. First 10 columns in first row: ");
        for(int i = 0; i < 10; i++){
            if(i < 9){
                System.out.printf(" %.2f , ", matrixC.get(0,i));
            }
            else{
                System.out.printf(" %.2f %n", matrixC.get(0,i));
            }
        }
        
        System.out.printf("product completed in %.2f seconds.%n", (endTime - startTime) / 1e9);
    }

    static class MatrixMultiplier implements Runnable {
        int matrixSize, startRow, endRow, blockSize;
        DoubleMatrix blockMatrixA;
        DoubleMatrix blockMatrixB;
        DoubleMatrix blockProduct;

        MatrixMultiplier(int matrixSize, int startRow, int endRow) {
            
            this.matrixSize = matrixSize;
            this.startRow = startRow;
            this.endRow = endRow;
            this.blockSize = endRow - startRow;
            this.blockMatrixA = new DoubleMatrix(blockSize, matrixSize);
            this.blockMatrixB = new DoubleMatrix(matrixSize, blockSize);
            this.blockProduct = new DoubleMatrix(blockSize, blockSize);
        }

        public void getBlockMatrixA(){
            for(int i = startRow; i < endRow; i++){

                this.blockMatrixA.putRow(i - startRow, matrixA.getRow(i));
            }
        }

        public void getBlockMatrixB(int startColumn, int endColumn){

            for(int i = startColumn; i < endColumn; i++){

                this.blockMatrixB.putColumn(i - startColumn, matrixB.getColumn(i));
                }
        }

        public void setMatrixResultBlock(int startColumn, int endColumn){
            
            int rows = this.blockMatrixA.rows;

            for(int i = 0; i < rows; i++){
                for(int j = startColumn; j < endColumn; j++){
                    matrixC.put(startRow + i, j, this.blockProduct.get(i,j - startColumn));
                }
            }
        }

        @Override
        public void run() {

            getBlockMatrixA();

            for(int i = 0; i < matrixSize; i += this.blockSize){
                getBlockMatrixB(i, i +this.blockSize);
                this.blockProduct = this.blockMatrixA.mmul(this.blockMatrixB);
                setMatrixResultBlock(i, i + this.blockSize);
            }
        }
    }
}
